packages/amqp/lib/AbstractAmqpConsumer.ts

```diff
@@ -128,37 +133,47 @@ export abstract class AbstractAmqpConsumer<
         this.handleMessageProcessed(null, 'invalid_message', messageId.result)
         return
       }
+      const { originalMessage, parsedMessage } = deserializedMessage.result
+
       // @ts-ignore
-      const messageType = deserializedMessage.result[this.messageTypeField]
+      const messageType = parsedMessage[this.messageTypeField]
       const transactionSpanId = `queue_${this.queueName}:${
         // @ts-ignore
         // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
-        deserializedMessage.result[this.messageTypeField]
+        parsedMessage[this.messageTypeField]
       }`
 
       // @ts-ignore
-      const uniqueTransactionKey = deserializedMessage.result[this.messageIdField]
+      const uniqueTransactionKey = parsedMessage[this.messageIdField]
       this.transactionObservabilityManager?.start(transactionSpanId, uniqueTransactionKey)
       if (this.logMessages) {
-        const resolvedLogMessage = this.resolveMessageLog(deserializedMessage.result, messageType)
+        const resolvedLogMessage = this.resolveMessageLog(parsedMessage, messageType)
         this.logMessage(resolvedLogMessage)
       }
-      this.internalProcessMessage(deserializedMessage.result, messageType)
+      this.internalProcessMessage(parsedMessage, messageType)
         .then((result) => {
-          if (result.error === 'retryLater') {
-            this.channel.nack(message, false, true)
-            this.handleMessageProcessed(deserializedMessage.result, 'retryLater')
-          }
           if (result.result === 'success') {
             this.channel.ack(message)
-            this.handleMessageProcessed(deserializedMessage.result, 'consumed')
+            this.handleMessageProcessed(originalMessage, 'consumed')
+            return
+          }
+
+          // retryLater
+          const timestamp = this.tryToExtractTimestamp(originalMessage) ?? new Date()
+          // requeue the message if maxRetryDuration is not exceeded, else ack it to avoid infinite loop
+          if (!isRetryDateExceeded(timestamp, this.maxRetryDuration)) {
+            this.channel.nack(message, false, true)
+            this.handleMessageProcessed(originalMessage, 'retryLater')
+          } else {
+            this.channel.ack(message)
+            this.handleMessageProcessed(originalMessage, 'error')
```

Comment:
can we add a ToDo marker here to add DLQ invocation after we support it for RabbitMQ?

Reply:
Good point! already added 🙌